code/modules/power/apc.dm

```diff
@@ -909,104 +901,66 @@
 	else
 		return locked
 
-/obj/machinery/power/apc/Topic(href, href_list, var/usingUI = 1)
-	if(..())
-		return 1
-
-	if(!can_use(usr, 1))
-		return 1
-
-	if(href_list["lock"])
-		if(!is_authenticated(usr))
-			return
-
-		coverlocked = !coverlocked
-
-	else if(href_list["breaker"])
-		if(!is_authenticated(usr))
-			return
-
-		toggle_breaker()
-
-	else if(href_list["toggle_nightshift"])
-		if(!is_authenticated(usr))
-			return
-
-		if(last_nightshift_switch > world.time + 100) // don't spam...
-			to_chat(usr, "<span class='warning'>[src]'s night lighting circuit breaker is still cycling!</span>")
-			return
-		last_nightshift_switch = world.time
-		set_nightshift(!nightshift_lights)
-
-	else if(href_list["cmode"])
-		if(!is_authenticated(usr))
-			return
-
-		chargemode = !chargemode
-		if(!chargemode)
-			charging = 0
-			update_icon()
-
-	else if(href_list["eqp"])
-		if(!is_authenticated(usr))
-			return
-
-		var/val = text2num(href_list["eqp"])
-		equipment = setsubsystem(val)
-		update_icon()
-		update()
-
-	else if(href_list["lgt"])
-		if(!is_authenticated(usr))
-			return
-
-		var/val = text2num(href_list["lgt"])
-		lighting = setsubsystem(val)
-		update_icon()
-		update()
-
-	else if(href_list["env"])
-		if(!is_authenticated(usr))
-			return
-
-		var/val = text2num(href_list["env"])
-		environ = setsubsystem(val)
-		update_icon()
-		update()
-	else if( href_list["close"] )
-		SSnanoui.close_user_uis(usr, src)
-
-		return 0
-	else if(href_list["close2"])
-		usr << browse(null, "window=apcwires")
-
-		return 0
-
-	else if(href_list["overload"])
-		if(issilicon(usr) && !aidisabled)
-			overload_lighting()
-
-	else if(href_list["malfhack"])
-		if(get_malf_status(usr))
-			malfhack(usr)
-
-	else if(href_list["occupyapc"])
-		if(get_malf_status(usr))
-			malfoccupy(usr)
-
-	else if(href_list["deoccupyapc"])
-		if(get_malf_status(usr))
-			malfvacate()
-
-	else if(href_list["toggleaccess"])
-		if(istype(usr, /mob/living/silicon))
-			if(emagged || aidisabled || (stat & (BROKEN|MAINT)))
-				to_chat(usr, "The APC does not respond to the command.")
+/obj/machinery/power/apc/tgui_act(action, params)
+	if(..() || !can_use(usr, 1) || (locked && !usr.has_unlimited_silicon_privilege && action != "toggle_nightshift" && !usr.can_admin_interact()))
```

Comment:
`!can_use(usr, TRUE)`

Also I would encase `action != "toggle_nightshift"` in braces to aid readability